File: python/heap.py
import math

class Heap:
    def __init__(self, new_arr = None):
        self.arr = [None]*101
        if new_arr:
            index = 1
            for i in new_arr:
                self.arr[index] = i
                index += 1
            self.size = index - 1
        else:
            self.size = 0
       

    def insert(self, value):
        if self.size == 100:
            return
        self.size += 1
        self.arr[self.size] = value
        self.sift_up(self.size)

    def sift_up(self, index):
        while self.arr[self.parent(index)] < self.arr[index]:
            temp = self.arr[index]
            self.arr[index] = self.arr[self.parent(index)]
            self.arr[self.parent(index)] = temp
            index = self.parent(index)
    
    def max_heapify(self, index):
        left = self.left(index)
        right = self.right(index)
        largest = index

        if left <= self.size and self.arr[left] > self.arr[largest]:
            largest = left

        if right <= self.size and self.arr[right] > self.arr[largest]:
            largest = right

        if largest != index:
            temp = self.arr[index]
            self.arr[index] = self.arr[largest]
            self.arr[largest] = temp
            self.max_heapify(largest)

    def extract_max(self):
        if self.size == 0:
            raise Exception("Cannot extract from an empty heap")
        retVal = self.arr[1]
        self.size -= 1
        if self.size > 0:
            self.arr[1] = self.arr[self.size + 1]
            self.max_heapify(1)
        return retVal

    def parent(self, index):
        if index == 1:
            return 1
        else:
            return math.floor(index/2)

    def left(self, index):
        return 2 * index

    def right(self, index):
        return 2 * index + 1

    def get_size(self):
        return self.size

    def is_empty(self):
        return self.size == 0

    def remove(self, index):
        self.arr[index] = self.arr[self.size]
        self.size -= 1
        self.max_heapify(index)

    def build_max_heap(self):
        for i in range(math.floor(self.size/2), 0):
            self.max_heapify(i)

    def heap_sort(self):
        self.build_max_heap()
        end = self.size
        while end > 1:
            temp = self.arr[1]
            self.arr[1] = self.arr[end]
            self.arr[end] = temp
            end -= 1
            self.size -= 1
            self.max_heapify(1)

    def print_heap(self):
        print("Printing heap:")
        for i in range(1, self.size + 1):
            if i is not None:
                print(self.arr[i])
        print("----------------------")

    def get_arr(self):
        return self.arr

def main():
    heap = Heap()
    assert heap.is_empty()
    assert 0 == heap.get_size()

    # test insertion and subsequent max heap
    heap.insert(1)
    heap.insert(2)
    heap.insert(3)
    heap.insert(4)
    heap.insert(5)
    heap.insert(6)
    heap.insert(7)

    assert not heap.is_empty()
    assert 7 == heap.get_size()
    heap.print_heap()

    assert 7 == heap.extract_max()
    assert 6 == heap.get_size()
    assert 6 == heap.extract_max()
    assert 5 == heap.get_size()
    assert 5 == heap.extract_max()
    assert 4 == heap.get_size()
    assert 4 == heap.extract_max()
    assert 3 == heap.get_size()
    assert 3 == heap.extract_max()
    assert 2 == heap.get_size()
    assert 2 == heap.extract_max()
    assert 1 == heap.get_size()
    assert 1 == heap.extract_max()
    assert 0 == heap.get_size()
    assert heap.is_empty()

    heap.insert(1)
    heap.insert(2)
    heap.insert(3)
    heap.insert(4)
    heap.insert(5)
    heap.insert(6)
    heap.insert(7)
    heap.print_heap()

    heap.remove(4)
    heap.print_heap()
    heap.remove(7)
    heap.print_heap()
    heap.remove(1)
    heap.print_heap()

    # test build max heap on already created array
    newArr = []
    for i in range(100, 0, -1):
        newArr.append(i)
    
    heap2 = Heap(newArr)
    heap2.print_heap()

    assert not heap2.is_empty()
    assert 100 == heap2.get_size()
    heap2.build_max_heap()
    heap2.print_heap()
    assert 100 == heap2.extract_max()
    assert 99 == heap2.extract_max()
    assert 98 == heap2.extract_max()
    assert 97 == heap2.extract_max()
    assert 96 == heap2.extract_max()
    assert 95 == heap2.extract_max()
    assert 94 == heap2.extract_max()
    assert 93 == heap2.extract_max()
    assert 92 == heap2.extract_max()
    assert 91 == heap2.extract_max()
    assert 90 == heap2.extract_max()
    assert 89 == heap2.extract_max()
    assert 88 == heap2.extract_max()
    assert 87 == heap2.extract_max()
    assert 86 == heap2.extract_max()
    assert 85 == heap2.extract_max()
    assert 84 == heap2.extract_max()
    assert 83 == heap2.extract_max()
    assert 82 == heap2.extract_max()
    assert 81 == heap2.extract_max()
    assert 80 == heap2.extract_max()
    assert 79 == heap2.extract_max()
    assert 78 == heap2.extract_max()
    assert 77 == heap2.extract_max()
    assert 76 == heap2.extract_max()
    assert 75 == heap2.extract_max()
    assert 74 == heap2.extract_max()
    assert 73 == heap2.extract_max()
    assert 72 == heap2.extract_max()
    assert 71 == heap2.extract_max()
    assert 70 == heap2.extract_max()
    assert 69 == heap2.extract_max()
    assert 68 == heap2.extract_max()
    assert 67 == heap2.extract_max()
    assert 66 == heap2.extract_max()
    assert 65 == heap2.extract_max()
    assert 64 == heap2.extract_max()
    assert 63 == heap2.extract_max()
    assert 62 == heap2.extract_max()
    assert 61 == heap2.extract_max()
    assert 60 == heap2.extract_max()
    assert 59 == heap2.extract_max()
    assert 58 == heap2.extract_max()
    assert 57 == heap2.extract_max()
    assert 56 == heap2.extract_max()
    assert 55 == heap2.extract_max()
    assert 54 == heap2.extract_max()
    assert 53 == heap2.extract_max()
    assert 52 == heap2.extract_max()
    assert 51 == heap2.extract_max()
    assert 50 == heap2.extract_max()
    assert 49 == heap2.extract_max()
    assert 48 == heap2.extract_max()
    assert 47 == heap2.extract_max()
    assert 46 == heap2.extract_max()
    assert 45 == heap2.extract_max()
    assert 44 == heap2.extract_max()
    assert 43 == heap2.extract_max()
    assert 42 == heap2.extract_max()
    assert 41 == heap2.extract_max()
    assert 40 == heap2.extract_max()
    assert 39 == heap2.extract_max()
    assert 38 == heap2.extract_max()
    assert 37 == heap2.extract_max()
    assert 36 == heap2.extract_max()
    assert 35 == heap2.extract_max()
    assert 34 == heap2.extract_max()
    assert 33 == heap2.extract_max()
    assert 32 == heap2.extract_max()
    assert 31 == heap2.extract_max()
    assert 30 == heap2.extract_max()
    assert 29 == heap2.extract_max()
    assert 28 == heap2.extract_max()
    assert 27 == heap2.extract_max()
    assert 26 == heap2.extract_max()
    assert 25 == heap2.extract_max()
    assert 24 == heap2.extract_max()
    assert 23 == heap2.extract_max()
    assert 22 == heap2.extract_max()
    assert 21 == heap2.extract_max()
    assert 20 == heap2.extract_max()
    assert 19 == heap2.extract_max()
    assert 18 == heap2.extract_max()
    assert 17 == heap2.extract_max()
    assert 16 == heap2.extract_max()
    heap2.print_heap()
    assert 15 == heap2.extract_max()
    assert 14 == heap2.extract_max()
    assert 13 == heap2.extract_max()
    assert 12 == heap2.extract_max()
    assert 11 == heap2.extract_max()
    assert 10 == heap2.extract_max()
    assert 9 == heap2.extract_max()
    assert 8 == heap2.extract_max()
    assert 7 == heap2.extract_max()
    assert 6 == heap2.extract_max()
    assert 5 == heap2.extract_max()
    assert 4 == heap2.extract_max()
    heap2.print_heap()
    assert 3 == heap2.extract_max()
    assert 2 == heap2.extract_max()
    assert 1 == heap2.extract_max()
        
    # test heapsort on already created array
    newArr2 = []
    for i in range(100, 0, -1):
        newArr2.append(i)
    
    heap3 = Heap(newArr2)
    assert not heap3.is_empty()

    heap3.heap_sort()
    vec = heap3.get_arr()
    print("Printing sorted array:")
    for i in vec:
        if i is not None:
            print(i)
    print("----------------------")
    for i in range(1, 101):
        assert i == vec[i]

if __name__ == "__main__":
    main()

File: java/Heap.java
public class Heap {
    private int[] arr;
    private int size;
    
    public Heap() {
        this.arr = new int[101];
        this.size = 0;
    }

    public Heap(int[] arr) {
        this.arr = new int[101];
        this.size = 0;
        int index = 1;
        for (int i : arr) {
            this.arr[index] = i;
            index++;
        }
        this.size = index - 1;
    }

    public boolean is_empty() {
        return this.size == 0;
    }

    public int get_size() {
        return this.size;
    }

    public void insert(int value) {
        if (this.size == 100) return;
        this.size++;
        this.arr[this.size] = value;
        sift_up(this.size);
    }

    public void sift_up(int index) {
        while (this.arr[parent(index)] < this.arr[index]) {
            int temp = this.arr[parent(index)];
            this.arr[parent(index)] = this.arr[index];
            this.arr[index] = temp;
            index = parent(index);
        }
    }

    private int parent(int index) {
        if (index == 1) return 1;
        else return (int) Math.floor(index/2);
    }

    private int left(int index) {
        return index * 2;
    }

    private int right(int index) {
        return index * 2 + 1;
    }

    public int extract_max() {
        if (this.size == 0) {
            throw new RuntimeException();
        }
        int retVal = this.arr[1];
        this.size--;
        if (size > 0) {
            this.arr[1] = this.arr[size+1];
            max_heapify(1);
        }
        return retVal;
    }

    public void max_heapify(int index) {
        int largest = index;
        int left = left(index);
        int right = right(index);
        if (left <= this.size && this.arr[left] > this.arr[largest]) {
            largest = left;
        }
        if (right <= this.size && this.arr[right] > this.arr[largest]) {
            largest = right;
        }
        if (largest != index) {
            int temp = this.arr[index];
            this.arr[index] = this.arr[largest];
            this.arr[largest] = temp;
            max_heapify(largest);
        }
    }

    public int[] get_arr() {
        return this.arr;
    }

    public void print_heap() {
        System.out.println("Printing heap:");
        for (int i = 1; i <= this.size; i++) {
            System.out.println(this.arr[i]);
        }
        System.out.println("----------------------");
        
    }

    public void remove(int index) {
        this.arr[index] = this.arr[this.size];
        this.size--;
        max_heapify(index);
    }

    public void build_max_heap() {
        for(int i = (int) Math.floor(this.size/2); i >= 1; i--) {
            max_heapify(i);
        }
    }

    public void heap_sort() {
        build_max_heap();
        int end = this.size;
        while (end > 1) {
            int temp = this.arr[1];
            this.arr[1] = this.arr[end];
            this.arr[end] = temp;
            this.size--;
            max_heapify(1);
            end--;
        }
    }

    public static void asserting(boolean assertion) {
        if (!assertion) throw new RuntimeException();
    }

    public static void main(String[] args) {
        Heap heap = new Heap();
        asserting(heap.is_empty());
        asserting(0 == heap.get_size());

        // test insertion and subsequent max heap
        heap.insert(1);
        heap.insert(2);
        heap.insert(3);
        heap.insert(4);
        heap.insert(5);
        heap.insert(6);
        heap.insert(7);

        asserting(!heap.is_empty());
        asserting(7 == heap.get_size());
        heap.print_heap();

        asserting(7 == heap.extract_max());
        asserting(6 == heap.get_size());
        asserting(6 == heap.extract_max());
        asserting(5 == heap.get_size());
        asserting(5 == heap.extract_max());
        asserting(4 == heap.get_size());
        asserting(4 == heap.extract_max());
        asserting(3 == heap.get_size());
        asserting(3 == heap.extract_max());
        asserting(2 == heap.get_size());
        asserting(2 == heap.extract_max());
        asserting(1 == heap.get_size());
        asserting(1 == heap.extract_max());
        asserting(0 == heap.get_size());
        asserting(heap.is_empty());

        heap.insert(1);
        heap.insert(2);
        heap.insert(3);
        heap.insert(4);
        heap.insert(5);
        heap.insert(6);
        heap.insert(7);
        heap.print_heap();

        heap.remove(4);
        heap.print_heap();
        heap.remove(7);
        heap.print_heap();
        heap.remove(1);
        heap.print_heap();

        // test build max heap on already created array
        int[] newArr = new int[100];
        int index = 0;
        for (int i = 100; i >= 1; i--) {
            newArr[index] = i;
            index++;
        }
        Heap heap2 = new Heap(newArr);

        asserting(!heap2.is_empty());
        asserting(100 == heap2.get_size());
        heap2.build_max_heap();
        heap2.print_heap();
        asserting(100 == heap2.extract_max());
        asserting(99 == heap2.extract_max());
        asserting(98 == heap2.extract_max());
        asserting(97 == heap2.extract_max());
        asserting(96 == heap2.extract_max());
        asserting(95 == heap2.extract_max());
        asserting(94 == heap2.extract_max());
        asserting(93 == heap2.extract_max());
        asserting(92 == heap2.extract_max());
        asserting(91 == heap2.extract_max());
        asserting(90 == heap2.extract_max());
        asserting(89 == heap2.extract_max());
        asserting(88 == heap2.extract_max());
        asserting(87 == heap2.extract_max());
        asserting(86 == heap2.extract_max());
        asserting(85 == heap2.extract_max());
        asserting(84 == heap2.extract_max());
        asserting(83 == heap2.extract_max());
        asserting(82 == heap2.extract_max());
        asserting(81 == heap2.extract_max());
        asserting(80 == heap2.extract_max());
        asserting(79 == heap2.extract_max());
        asserting(78 == heap2.extract_max());
        asserting(77 == heap2.extract_max());
        asserting(76 == heap2.extract_max());
        asserting(75 == heap2.extract_max());
        asserting(74 == heap2.extract_max());
        asserting(73 == heap2.extract_max());
        asserting(72 == heap2.extract_max());
        asserting(71 == heap2.extract_max());
        asserting(70 == heap2.extract_max());
        asserting(69 == heap2.extract_max());
        asserting(68 == heap2.extract_max());
        asserting(67 == heap2.extract_max());
        asserting(66 == heap2.extract_max());
        asserting(65 == heap2.extract_max());
        asserting(64 == heap2.extract_max());
        asserting(63 == heap2.extract_max());
        asserting(62 == heap2.extract_max());
        asserting(61 == heap2.extract_max());
        asserting(60 == heap2.extract_max());
        asserting(59 == heap2.extract_max());
        asserting(58 == heap2.extract_max());
        asserting(57 == heap2.extract_max());
        asserting(56 == heap2.extract_max());
        asserting(55 == heap2.extract_max());
        asserting(54 == heap2.extract_max());
        asserting(53 == heap2.extract_max());
        asserting(52 == heap2.extract_max());
        asserting(51 == heap2.extract_max());
        asserting(50 == heap2.extract_max());
        asserting(49 == heap2.extract_max());
        asserting(48 == heap2.extract_max());
        asserting(47 == heap2.extract_max());
        asserting(46 == heap2.extract_max());
        asserting(45 == heap2.extract_max());
        asserting(44 == heap2.extract_max());
        asserting(43 == heap2.extract_max());
        asserting(42 == heap2.extract_max());
        asserting(41 == heap2.extract_max());
        asserting(40 == heap2.extract_max());
        asserting(39 == heap2.extract_max());
        asserting(38 == heap2.extract_max());
        asserting(37 == heap2.extract_max());
        asserting(36 == heap2.extract_max());
        asserting(35 == heap2.extract_max());
        asserting(34 == heap2.extract_max());
        asserting(33 == heap2.extract_max());
        asserting(32 == heap2.extract_max());
        asserting(31 == heap2.extract_max());
        asserting(30 == heap2.extract_max());
        asserting(29 == heap2.extract_max());
        asserting(28 == heap2.extract_max());
        asserting(27 == heap2.extract_max());
        asserting(26 == heap2.extract_max());
        asserting(25 == heap2.extract_max());
        asserting(24 == heap2.extract_max());
        asserting(23 == heap2.extract_max());
        asserting(22 == heap2.extract_max());
        asserting(21 == heap2.extract_max());
        asserting(20 == heap2.extract_max());
        asserting(19 == heap2.extract_max());
        asserting(18 == heap2.extract_max());
        asserting(17 == heap2.extract_max());
        asserting(16 == heap2.extract_max());
        heap2.print_heap();
        asserting(15 == heap2.extract_max());
        asserting(14 == heap2.extract_max());
        asserting(13 == heap2.extract_max());
        asserting(12 == heap2.extract_max());
        asserting(11 == heap2.extract_max());
        asserting(10 == heap2.extract_max());
        asserting(9 == heap2.extract_max());
        asserting(8 == heap2.extract_max());
        asserting(7 == heap2.extract_max());
        asserting(6 == heap2.extract_max());
        asserting(5 == heap2.extract_max());
        asserting(4 == heap2.extract_max());
        asserting(3 == heap2.extract_max());
        asserting(2 == heap2.extract_max());
        asserting(1 == heap2.extract_max());
            
        // test heapsort on already created array
        int[] newArr2 = new int[100];
        index = 0;
        for (int i = 100; i >= 1; i--) {
            newArr2[index] = i;
            index++;
        }
        Heap heap3 = new Heap(newArr2);
        asserting(!heap3.is_empty());

        heap3.heap_sort();
        int[] vec = heap3.get_arr();
        System.out.println("Printing sorted array:");
        for (int i = 1; i <= 100; i++) {
            System.out.println(vec[i]);
        }
        System.out.println("----------------------");
        for (int i = 1; i <= 100; i++) {
            asserting(i == vec[i]);
        }
    }
}